### PR TITLE
feat: Add chat template customization to ilab serve

### PR DESCRIPTION
**Pull request by n1hility**
Introduces a `--chat-template` option, covering both vllm and llama.cpp backends. This addresses compatibility problems with ilab chat clients and the vllm backend, and brings closer consitency between llama.cpp and vllm for out of the box usage. This is also a prerequisite for e2e testing using vLLM.

The option has three possible values:

1. 'auto' (default) - instructlab overrides the model template with a new template based on hard coded model knowledege. This is the current behavior before this commit. However, in the future this could be changed to a different dynamic policy, as the name simply implies instructlab decides

2. 'tokenizer' - uses the template expressed in the model's tokenizer_config.json file (or internal gguf settings). This is the ideal long term direction, although 'auto' allows for better compatibility with legacy behvior.

3. A filesystem path expressing the location of a custom template (e.g. /foo/bar/template.jinja)

Notes:

- On vLLM, Temporary files are held in globals with an atexit cleanup since we currently lack a well defined scope with a long enough period to cover vLLM opening the file. A future change could address this by introducing a scope, or shift startup wait code into start.
- Even though the effective default of --chat-template is "auto" internally None is used to allow for the usage of context overrides.
- Fixes leftover instance logger in exception handling
- Testing is mostly a follow-up with the e2e fixes
- Will link dev doc update soon

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the
  [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [x] [Changelog](https://github.com/instructlab/instructlab/blob/main/CHANGELOG.md) updated with breaking and/or notable changes for the next minor release.
- [x] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [x] functional tests have been added, if necessary.

### CODE REVIEW

**Comment by n1hility** [comment]
@alimaredia @leseb @danmcp PTAL

**Comment by danmcp** [comment]
Unit tests would be nice to have here.

**Comment by n1hility** [comment]
> Unit tests would be nice to have here.

+1 Added some additional tests for chat-template values on top of the config test

**Comment by n1hility** [comment]
Added hold for the dev doc portion

File: src/instructlab/model/backends/backends.py
Comment (leseb) [inline src/instructlab/model/backends/backends.py:356]: I think this can be avoided by simply setting a default to chat_template (as suggested above).

File: src/instructlab/model/backends/llama_cpp.py
Comment (leseb) [inline src/instructlab/model/backends/llama_cpp.py:205]: Can you add a comment indicating that this third condition means `chat_template` is a template file?

File: src/instructlab/model/backends/llama_cpp.py
Comment (leseb) [inline src/instructlab/model/backends/llama_cpp.py:250]: What if `template_path` is a directory?

File: src/instructlab/model/backends/llama_cpp.py
Comment (leseb) [inline src/instructlab/model/backends/llama_cpp.py:255]: Have do we handle `path.read_text` failing?

File: src/instructlab/model/backends/vllm.py
Comment (leseb) [inline src/instructlab/model/backends/vllm.py:58]: How is this related to `chat_template`?

File: src/instructlab/model/backends/llama_cpp.py
Comment (leseb) [reply src/instructlab/model/backends/llama_cpp.py:205]: Also, this conditional block serves as a validation for `chat_template` should we extract it in a `validate_chat_template` and call it earlier (to fail early if needed)?

File: src/instructlab/model/backends/vllm.py
Comment (leseb) [inline src/instructlab/model/backends/vllm.py:146]: Should we use a jinja lib to validate this template before sending it to vllm? Or build using a jinja lib maybe?

File: src/instructlab/model/backends/vllm.py
Comment (leseb) [inline src/instructlab/model/backends/vllm.py:204]: Leftover or intentional?

File: src/instructlab/model/backends/vllm.py
Comment (leseb) [inline src/instructlab/model/backends/vllm.py:206]: Let's comment that this is the case where a file is passed for chat_template

File: src/instructlab/model/serve.py
Comment (leseb) [inline src/instructlab/model/serve.py:64]:
```suggestion
        "'tokenizer', or a path to jinja formatted template file. \n"
```

It relies on jinja right?

File: tests/test_lab_serve.py
Comment (leseb) [inline tests/test_lab_serve.py:84]:
```suggestion
def test_chat_custom(mock_popen, mock_determine, tmp_path: pathlib.Path, _):
    file = tmp_path / "chat_template.jinja2"
    args = vllm_setup_test(
        mock_popen,
        mock_determine,
        [
            "--config=DEFAULT",
            "model",
            "serve",
            "--model-path=foo",
            "--chat-template={}".format(file.name),
        ],
    )
    assert_vllm_args(args)
    assert_template(
        args=args, expect_chat=True, path_chat=True, chat_value=file.name
    )
```

This should work. @tiran pointed out in another review that using the `tmp_path` fixture is more appropriate.

**Review by leseb — CHANGES_REQUESTED** [review]
Nice work! Can you look into adding functional tests (`scripts/functional-tests.sh`) as well to at least validate this with llama-cpp? Thanks

File: src/instructlab/model/backends/llama_cpp.py
Comment (n1hility) [reply src/instructlab/model/backends/llama_cpp.py:255]: So IMO this will throw appropriate errors with the path (FileNotFound, OSError etc). Open to ideas if you have a better suggestion.

However this is a problem were we dont handle these outer sections correctly will fix that.

File: src/instructlab/model/backends/vllm.py
Comment (n1hility) [reply src/instructlab/model/backends/vllm.py:58]: The Auto handler requires the use of temp files that we need to cleanup, yet still stick around long enough for vLLM to open and read them (see the first bullet in the notes section for further details)

File: src/instructlab/model/backends/vllm.py
Comment (n1hility) [reply src/instructlab/model/backends/vllm.py:204]: Oops! good catch

File: src/instructlab/configuration.py
Comment (n1hility) [reply src/instructlab/configuration.py:176]: Thanks for the feedback. So the intention was to be able to detect and allow overrides like some of the other values are setup for but happy to go with a hard constant if you like that better.

File: src/instructlab/model/backends/llama_cpp.py
Comment (n1hility) [reply src/instructlab/model/backends/llama_cpp.py:205]: Since both vllm and llama cpp log the chat template and error properly I preferred just delegating to them and avoid the double parsing.

File: src/instructlab/model/backends/vllm.py
Comment (n1hility) [reply src/instructlab/model/backends/vllm.py:146]: (see above on that one, assuming I understood your comment correctly)

File: tests/test_lab_serve.py
Comment (n1hility) [reply tests/test_lab_serve.py:84]: thanks good suggestion

File: src/instructlab/model/backends/llama_cpp.py
Comment (leseb) [reply src/instructlab/model/backends/llama_cpp.py:205]: Souns good, just a small note about the third case 🙏🏻 so it's easier to understand for whoever is reading at the code and not familiar with it.

File: src/instructlab/model/backends/vllm.py
Comment (leseb) [reply src/instructlab/model/backends/vllm.py:58]: Ahhhh ok ok :) Thanks for the clarification.

**Comment by tiran** [comment]
The temporary file handling looks complicated. It's going to be a nightmare the write unit and function tests for its behavior.

I wonder if it would be possible to write this differently:
- extend `BackendServer` ABC with a `tempdir` attribute
- add a method to initialize `tempdir`
- cleanup and remove `tempdir` in `shutdown` / exit route
- convert the template functions to methods of the BackendServer class and the subclasses.

This would bind the life time of the temporary directory to the life time of the server.

**Comment by alimaredia** [comment]
@tiran @n1hility Christian touched on what I had been thinking when I read through the code. I had never heard of `atexit` before this PR so I would have expected the temp file cleanup to happen in `shutdown()` and I think the changes he proposes are easier to read. 

One other thing to point out is that `--model-family` is an argument/config for `ilab model chat` https://github.com/instructlab/instructlab/blob/main/src/instructlab/model/chat.py#L185 and it needs to be set now in the vLLM section of the serve config, not just for llama_cpp.

**Comment by leseb** [comment]
@tiran @alimaredia @n1hility the vLLM [doc](https://docs.vllm.ai/en/latest/serving/openai_compatible_server.html#chat-template) says:

> For those model, you can manually specify their chat template in the --chat-template parameter with the file path to the chat template, or the template in string form

Also, the [code](https://github.com/vllm-project/vllm/blob/3963a5335bb4106f2ecd1139527e3568d2151933/vllm/entrypoints/openai/serving_chat.py#L71-L89) seems to fallback if the file cannot be read to decoding the string.

If we use that approach we don't need to create a temp file anymore and just pass the template as a single line which simplifies the implementation quite a lot.
Thoughts?

**Comment by n1hility** [comment]
> The temporary file handling looks complicated. It's going to be a nightmare the write unit and function tests for its behavior.
> 
> I wonder if it would be possible to write this differently:
> 
> * extend `BackendServer` ABC with a `tempdir` attribute
> * add a method to initialize `tempdir`
> * cleanup and remove `tempdir` in `shutdown` / exit route
> * convert the template functions to methods of the BackendServer class and the subclasses.
> 
> This would bind the life time of the temporary directory to the life time of the server.

Thanks for the feedback! This is what I was referring to in the first bullet of notes. Something like this would be be the ideal way to go. The problem was the scoping / lifecycle wasn't complete in all cases so would have lead to a larger change or risked leaking files, so just went with the simpler slightly less ideal path. I can take another look though.

**Comment by leseb** [comment]
> > If we use that approach we don't need to create a temp file anymore and just pass the template as a single line which simplifies the implementation quite a lot.
> > Thoughts?
> 
> The reason we need the temp file is just to support the "auto" usage, where ilab is providing the template (it needs to be written to a file for it to be used)

But if ilab provides a file already we don't need to manage it, just read it, right?

**Comment by n1hility** [comment]
> > > If we use that approach we don't need to create a temp file anymore and just pass the template as a single line which simplifies the implementation quite a lot.
> > > Thoughts?
> > 
> > 
> > The reason we need the temp file is just to support the "auto" usage, where ilab is providing the template (it needs to be written to a file for it to be used)
> 
> But if ilab provides a file already we don't need to manage it, just read it, right?

Oh you mean provide the templates as data file as static content in the distribution, inside the python module or something like that?

**Comment by n1hility** [comment]
> > The temporary file handling looks complicated. It's going to be a nightmare the write unit and function tests for its behavior.
> > I wonder if it would be possible to write this differently:
> > 
> > * extend `BackendServer` ABC with a `tempdir` attribute
> > * add a method to initialize `tempdir`
> > * cleanup and remove `tempdir` in `shutdown` / exit route
> > * convert the template functions to methods of the BackendServer class and the subclasses.
> > 
> > This would bind the life time of the temporary directory to the life time of the server.
> 
> Thanks for the feedback! This is what I was referring to in the first bullet of notes. Something like this would be be the ideal way to go. The problem was the scoping / lifecycle wasn't complete in all cases so would have lead to a larger change or risked leaking files, so just went with the simpler slightly less ideal path. I can take another look though.

I audited all the exception code paths and with one small adjustment I think we can use the instance scope. Will update soon.

**Comment by leseb** [comment]
> > > > If we use that approach we don't need to create a temp file anymore and just pass the template as a single line which simplifies the implementation quite a lot.
> > > > Thoughts?
> > > 
> > > 
> > > The reason we need the temp file is just to support the "auto" usage, where ilab is providing the template (it needs to be written to a file for it to be used)
> > 
> > 
> > But if ilab provides a file already we don't need to manage it, just read it, right?
> 
> Oh you mean provide the templates as data file as static content in the distribution, inside the python module or something like that?

Yes.

**Comment by n1hility** [comment]
> > > > > If we use that approach we don't need to create a temp file anymore and just pass the template as a single line which simplifies the implementation quite a lot.
> > > > > Thoughts?
> > > > 
> > > > 
> > > > The reason we need the temp file is just to support the "auto" usage, where ilab is providing the template (it needs to be written to a file for it to be used)
> > > 
> > > 
> > > But if ilab provides a file already we don't need to manage it, just read it, right?
> > 
> > 
> > Oh you mean provide the templates as data file as static content in the distribution, inside the python module or something like that?
> 
> Yes.

Took a quick look and this option seemed to have a bit more complexity due to the formats being slightly different yet still primarily duplicative between vllm and llama cpp.  I was able to leverage the backend instance scope after all after auditing the exception call paths (mostly only needed on adjustment to enable that).

File: src/instructlab/model/backends/vllm.py
Comment (alimaredia) [inline src/instructlab/model/backends/vllm.py:174]: What's the purpose of having a list here? From what I can tell `tmp_files` is either empty or contains one entry.

File: src/instructlab/model/backends/vllm.py
Comment (alimaredia) [inline src/instructlab/model/backends/vllm.py:123]: Chat template and model family need to be added to the args list in this function annotation.

File: src/instructlab/model/backends/vllm.py
Comment (alimaredia) [inline src/instructlab/model/backends/vllm.py:197]: @n1hility Tell me if I've got this flow right. If `ilab model serve` is called without the `--chat-template` flag or any additional `vllm_args`, the value of `chat_template` that is passed into the `vllm.Server()` constructor is `None`. 

That `None` value flows down to this conditional and then the python call to vLLM is appended with `--chat-template None`. Is `None` a valid value for that flag in vLLM?

When I first added the vLLM serving code, I decided that the internal chat template selection via model family was not necessary for a first pass and that passing it via `vllm_args` would be enough since at least granite and merlinite served just fine without any additional chat template configuration. 

I just want to make sure that's the case and that just plane `ilab model serve` for granite and merlinite just work out of the box.

File: src/instructlab/model/serve.py
Comment (alimaredia) [inline src/instructlab/model/serve.py:62]: After looking at and addressing https://github.com/instructlab/instructlab/pull/1636/files#r1677217137, would we want to consider setting the default here to `tokenizer just so that `ilab model serve` with granite and merlinite out of the box just works`?

File: src/instructlab/model/backends/vllm.py
Comment (alimaredia) [inline src/instructlab/model/backends/vllm.py:197]: @n1hility Trying to think of edge cases, Would we want to to make sure the file passed as chat_template exists on the file system? If you pass in a chat template that does not exist to vLLM does it error out and not start up or give a warning of some sort?

 In the cases of `run_vLLM()` being run with background=True, since there is no vLLM output, a user of `ilab`  commands that serve with vLLM under the hood wouldn't see these warning or have vLLM fail in the background.
 
 Maybe we add a check here and some debug logging saying what the chat template file name is if it's set?

File: src/instructlab/model/backends/llama_cpp.py
Comment (danmcp) [inline src/instructlab/model/backends/llama_cpp.py:220]: Nit: Better written as  logger.info("Replacing chat template:\n %s", template)

File: src/instructlab/model/backends/vllm.py
Comment (danmcp) [inline src/instructlab/model/backends/vllm.py:213]: Nit: Better written as: logger.info("vLLM starting up on pid %s at %s", vllm_process.pid, api_base)

File: tests/test_lab_serve.py
Comment (danmcp) [inline tests/test_lab_serve.py:21]: Noting the different style @tiran developed: https://github.com/instructlab/instructlab/pull/1617/files

Probably doesn't need to be a blocker for this PR

**Review by danmcp — APPROVED** [review]
Approving with a couple nits/follow-ups

File: tests/test_lab_serve.py
Comment (n1hility) [reply tests/test_lab_serve.py:21]: ah thanks didnt notice this.

Comment (tiran) [reply tests/test_lab_serve.py:21]: I recommend to use the `cli_runner` fixture. It takes are of new storage locations, too. Otherwise you may leak test files into your personal `~/.local/share` directory tree.

File: src/instructlab/model/backends/vllm.py
Comment (n1hility) [reply src/instructlab/model/backends/vllm.py:174]: Its's more for the future, since any passing of a large value or hard to parse value on the prompt we would need to store to the fs

File: src/instructlab/model/backends/vllm.py
Comment (n1hility) [reply src/instructlab/model/backends/vllm.py:123]: good catch! thanks

File: src/instructlab/model/backends/vllm.py
Comment (n1hility) [reply src/instructlab/model/backends/vllm.py:197]: ah yes this is a good idea for the background case

Comment (alimaredia) [reply src/instructlab/model/backends/vllm.py:197]: Would you want that to happen for follow up after this PR is merged? I can create an issue, and Seb and I can work on getting that in over the next 24 hours.

File: src/instructlab/model/backends/vllm.py
Comment (leseb) [inline src/instructlab/model/backends/vllm.py:192]: I still don't understand why we need a file where vLLM supports to be given a string directly. @n1hility did I miss the reason somewhere?

**Review by leseb — COMMENTED** [review]
I just have one leftover question otherwise this feels ready to go :) Thanks!

File: src/instructlab/model/backends/vllm.py
Comment (n1hility) [reply src/instructlab/model/backends/vllm.py:192]: @leseb yeah the big reason is just that files as arguments run up against problems: parsing/escaping issues, ARG_MAX limits, and cut and paste problems (e..g message is printed unescaped, which won't work with paste, user fixes but gets it wrong, etc). 

In the case where the args are passed to exec(), escaping usually isn't a problem, however, it does rely on the receiver of the args to handle them correctly. For example if the called code use an argument parser which optionally splits on = or some other form of interpolation, that can easily break with a template string in a param. Or if there is some sort of intermediary (for example, they construct a command line which is passed as a shell string). Some OSes also requote (e.g. windows) which can lead to oddities.  

Some projects that do this sort of thing use base64 encoding to avoid the problems (with the exception of ARG_MAX limits)

File: tests/test_lab_serve.py
Comment (n1hility) [reply tests/test_lab_serve.py]: Very cool @tiran !

**Review by alimaredia — APPROVED** [review]

**Review by leseb — APPROVED** [review]

**Pull request merged**
